staging/src/k8s.io/apiserver/pkg/endpoints/request/auditid.go

```diff
@@ -0,0 +1,65 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package request
+
+import (
+	"context"
+	"net/http"
+
+	"k8s.io/apimachinery/pkg/types"
+	auditinternal "k8s.io/apiserver/pkg/apis/audit"
+)
+
+type auditIDKeyType int
+
+// auditIDKey is the key to associate the Audit-ID value of a request.
+const auditIDKey auditIDKeyType = iota
+
+// WithAuditID returns a copy of the parent context into which the Audit-ID
+// associated with the request is set.
+//
+// If the specified auditID is empty, no value is set and the parent context is returned as is.
+func WithAuditID(parent context.Context, auditID types.UID) context.Context {
+	if auditID == "" {
+		return parent
+	}
+	return WithValue(parent, auditIDKey, auditID)
+}
+
+// AuditIDFrom returns the value of the audit ID associated with a request.
+//
+// First, we check if the audit ID has been attached to the the request context.
+// If not, we retrieve it from the request writeToHeader.
+func AuditIDFrom(r *http.Request) (types.UID, bool) {
+	ctx := r.Context()
+	if auditID, ok := ctx.Value(auditIDKey).(types.UID); ok {
+		return auditID, ok
+	}
+
+	if auditID := r.Header.Get(auditinternal.HeaderAuditID); len(auditID) > 0 {
```

Comment:
I find all the logic starting from here irritating in this context. Would prefer exactly one place where we have header logic.

Reply:
@sttts This is an optimization - If the audit ID is in the request header and we can use it as is then avoid copying it to the context. do you want me to go back to `always copy the audit ID in the request context`?

Reply:
@sttts I removed the optimization, the context always has an audit ID.